csrc/scheduler/mma_utils.cpp

```diff
@@ -1210,143 +1210,122 @@ RolesMapOpt getTensorsRoles(
+      // Don't map TVs to roles if they have unmapped dims
+      continue;
+    }
+    if (has_m && has_k && !has_n) {
+      roles_map[MatmulRole::INPUT_A].push_back(tv);
+      continue;
+    }
+    if (has_n && has_k && !has_m) {
+      roles_map[MatmulRole::INPUT_B].push_back(tv);
+      continue;
+    }
+    // Bias vectors are assigned to INPUT_C role
+    if (!has_k) {
+      roles_map[MatmulRole::INPUT_C].push_back(tv);
+      continue;
+    }
+  }
+
+  std::vector<TensorView*> storage;
+  for (TensorView* tv : mma_output_candidates) {
+    bool has_m = false, has_n = false, has_k = false, has_unmapped = false;
+    for (IterDomain* id :
+         TensorDomain::noReductions(tv->getMaybeRFactorDomain())) {
+      if (id->isBroadcast()) {
+        // Ignore broadcasts in output
+        continue;
+      }
+      const ValGroup& g = exact_graph.toGroup(id);
+      auto it = group_to_domain.find(g);
+      if (it == group_to_domain.end()) {
+        // output tv has an unmapped non-broadcast dimension
+        has_unmapped = true;
         continue;
       }
+      has_m |= it->second == MatmulDomain::M;
+      has_n |= it->second == MatmulDomain::N;
+      has_k |= it->second == MatmulDomain::K;
     }
-
-    for (auto& [role, tvs] : roles_map) {
-      // NOTE: sort input roles in descending order by uses() size, and
-      //  if equal then by name() to ensure the stable ordering of tensor
-      //  views in collections assigned to the supported roles
-      std::sort(tvs.begin(), tvs.end(), [](TensorView* a, TensorView* b) {
-        return (a->uses().size() == b->uses().size())
-            ? (a->name() < b->name())
-            : (a->uses().size() > b->uses().size());
-      });
+    // NOTE: depending on fusion definition k domain may appear in the output:
+    //  - for mma_output == fusion output k domain is present
+    //  - for mma_output != fusion output (fusion with epilogue) k domain
+    //    is not present
+    if (has_k || has_unmapped) {
+      // Don't map TVs to output roles if they have unmapped dims, or if they
+      // have K dimension
+      continue;
     }
-  };
-
-  const auto findOutputRolesByDomains = [](const DependenciesMap& deps_map,
-                                           RolesMap& roles_map) {
-    std::vector<TensorView*> storage;
-    storage.reserve(deps_map.size());
 
-    for (const auto& entry : deps_map) {
-      const auto& domains = entry.second;
-      const auto begin = domains.begin();
-      const auto end = domains.end();
-
-      bool has_m = (end != std::find(begin, end, MatmulDomain::M));
-      bool has_n = (end != std::find(begin, end, MatmulDomain::N));
-
-      // NOTE: depending on fusion definition k domain may appear in the output:
-      //  - for mma_output == fusion output k domain is present
-      //  - for mma_output != fusion output (fusion with epilogue) k domain
-      //    is not present
+    // NOTE: the core fusion output tensors are the ones with m and n
```

Comment:
What do you mean by core fusion output tensors?

Reply:
That comment is left over from the previous implementation. It means that outputs we identify as D must have M and N. Others can possibly exist but are not given roles. Note that we can't currently reduce dimensions in the epilogue but that could be possible some day and we should revisit the notion of output roles at that point (see #2213).